src/eval.rs

```diff
@@ -50,19 +73,76 @@ pub enum ExprSource {
         from: ExprResultGc,
         index: ExprResultBox,
     },
+    /// `assert condition; body`
+    Assert {
+        /// the asserted condition
+        condition: ExprResultBox,
+        /// the body which is only evaluated if the assertion is true
+        body: ExprResultBox,
+    },
     /// Dynamic attribute, such as the curly braces in `foo.${toString (1+1)}`
     Dynamic {
         inner: ExprResultBox,
     },
+    /// `with inner; body`
+    With {
+        /// the expression used to provide bindings
+        inner: ExprResultGc,
+        /// the body evaluated with the new bindings
+        body: ExprResultBox,
+    },
     Ident {
         name: String,
     },
     Literal {
         value: NixValue,
     },
+    /// `if condition then body else else_body`
+    IfElse {
+        /// the condition evaluated
+        condition: ExprResultBox,
+        /// the body evaluated when the condition is true
+        body: ExprResultBox,
+        /// the body evaluated when the condition is false
+        else_body: ExprResultBox,
+    },
+    /// A string, possibly interpolated
+    String {
+        /// interpolated and literal parts of this string
+        parts: Vec<StringPartSource>,
+    },
+    /// `a.b or c`
+    OrDefault {
+        /// `a.b`, of type `Select`
+        index: ExprResultBox,
+        /// `c`
+        default: ExprResultBox,
+    },
     Paren {
         inner: ExprResultBox,
     },
+    /// `{ arg1, ... } @ args` in a lambda definition `{ arg1, ... } @ args: body`
+    Pattern {
+        /// for `{ arg1, arg2 ? default }: body`, a map `"arg1" => None, "arg2" => default`
+        entries: HashMap<String, Option<ExprResultGc>>,
+        /// whether the patter is incomplete (contains `...`)
+        ellipsis: bool,
+        /// the identifier bound by `@`
+        at: Option<String>,
```

Comment:
```suggestion
        at: Option<ExprResultGc>,
```

This would let us put the identifier in a Scope, and the Expression struct would also give us more info about where the identifier is located.

Reply:
I made it an ExprResultBox as there is no need for sharing as far as I can see